### PR TITLE
start of python3 migration

### PR DESCRIPTION
**Pull request by ratherDashing**
didn't get much time to test but it starts up just fine.  can someone field test this?

### CODE REVIEW

**Comment by deathbybandaid** [comment]
I think we need to do the linting prior to any real code changes

**Comment by deathbybandaid** [comment]
This looks like a good step forward, finally had a chance to glance at the code.

File: LocastService.py [hunk 1]
```diff
@@ -1,4 +1,4 @@
-import json, urllib2, time, os, sys, string
+import json, urllib.request, urllib.error, urllib.parse, time, os, sys, string
```
Comment (deathbybandaid) [inline LocastService.py:1]: For linting, imports should be on their own lines.

Comment (deathbybandaid) [inline LocastService.py:1]: For linting, we want each of these imports on their own lines

File: LocastService.py [hunk 2]
```diff
@@ -34,29 +34,29 @@ def login(self, username, password):
             return False
 
         # login
-        print("Logging into Locast using username " + username + "...")
+        print(("Logging into Locast using username " + username + "..."))
```
Comment (deathbybandaid) [inline LocastService.py:37]: Not sure why you added doubled-up parenthesis

File: LocastService.py [hunk 3]
```diff
@@ -34,29 +34,29 @@ def login(self, username, password):
...
             return False
-        except urllib2.HTTPError as httpError:
-            print("Error during login: " + str(httpError.reason))
+        except urllib.error.HTTPError as httpError:
+            print(("Error during login: " + str(httpError.reason)))
             return False
         except:
             loginErr = sys.exc_info()[0]
-            print("Error during login: " + loginErr.message)
+            print(("Error during login: " + loginErr.message))
```
Comment (deathbybandaid) [inline LocastService.py:59]: parenthesis

File: LocastService.py [hunk 4]
```diff
@@ -71,33 +71,33 @@ def validate_user(self):
...
             return False
-        except urllib2.HTTPError as httpError:
-            print("Error during user info request: " + str(httpError.reason))
+        except urllib.error.HTTPError as httpError:
+            print(("Error during user info request: " + str(httpError.reason)))
             return False
         except:
             userInfoErr = sys.exc_info()[0]
-            print("Error during user info request: " + userInfoErr.message)
+            print(("Error during user info request: " + userInfoErr.message))
```
Comment (deathbybandaid) [inline LocastService.py:88]: parenthesis

File: LocastService.py [hunk 5]
```diff
@@ -71,33 +71,33 @@ def validate_user(self):
...
             userOpn.close()
-        except urllib2.URLError as urlError:
-            print("Error during user info request: " + str(urlError.reason))
+        except urllib.error.URLError as urlError:
+            print(("Error during user info request: " + str(urlError.reason)))
             return False
-        except urllib2.HTTPError as httpError:
-            print("Error during user info request: " + str(httpError.reason))
+        except urllib.error.HTTPError as httpError:
+            print(("Error during user info request: " + str(httpError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:84]: parenthesis

File: LocastService.py [hunk 6]
```diff
@@ -71,33 +71,33 @@ def validate_user(self):
...
                                     headers={'Content-Type': 'application/json', 'authorization': 'Bearer ' + self.current_token})
 
-            userOpn = urllib2.urlopen(userReq)
+            userOpn = urllib.request.urlopen(userReq)
             userRes = json.load(userOpn)
             userOpn.close()
-        except urllib2.URLError as urlError:
-            print("Error during user info request: " + str(urlError.reason))
+        except urllib.error.URLError as urlError:
+            print(("Error during user info request: " + str(urlError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:81]: parenthesis

File: LocastService.py [hunk 7]
```diff
@@ -34,29 +34,29 @@ def login(self, username, password):
...
+                                              {'Content-Type': 'application/json'})
 
-            loginOpn = urllib2.urlopen(loginReq)
+            loginOpn = urllib.request.urlopen(loginReq)
             loginRes = json.load(loginOpn)
             loginOpn.close()
-        except urllib2.URLError as urlError:
-            print("Error during login: " + str(urlError.reason))
+        except urllib.error.URLError as urlError:
+            print(("Error during login: " + str(urlError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:52]: parenthesis

File: LocastService.py [hunk 8]
```diff
@@ -34,29 +34,29 @@ def login(self, username, password):
...
             return False
-        except urllib2.HTTPError as httpError:
-            print("Error during login: " + str(httpError.reason))
+        except urllib.error.HTTPError as httpError:
+            print(("Error during login: " + str(httpError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:55]: parenthesis

File: LocastService.py [hunk 9]
```diff
@@ -71,33 +71,33 @@ def validate_user(self):
...
         except:
             userInfoErr = sys.exc_info()[0]
-            print("Error during user info request: " + userInfoErr.message)
+            print(("Error during user info request: " + userInfoErr.message))
             return False
 
 
         print("User Info obtained.")
-        print("User didDonate: " + str(userRes['didDonate']))
+        print(("User didDonate: " + str(userRes['didDonate'])))
```
Comment (deathbybandaid) [inline LocastService.py:93]: parenthesis

File: LocastService.py [hunk 10]
```diff
@@ -71,33 +71,33 @@ def validate_user(self):
...
 
         # Check if donated
         if not userRes['didDonate']:
             print("Error!  User must donate for this to work.")
             return False
 
-        print("User donationExpire: " + str(userRes['donationExpire'] / 1000))
+        print(("User donationExpire: " + str(userRes['donationExpire'] / 1000)))
```
Comment (deathbybandaid) [inline LocastService.py:100]: parenthesis

File: LocastService.py [hunk 11]
```diff
@@ -111,21 +111,21 @@ def validate_user(self):
...
-                geoReq = urllib2.Request('https://get.geojs.io/v1/ip/geo.json')
-                geoOpn = urllib2.urlopen(geoReq)
+                geoReq = urllib.request.Request('https://get.geojs.io/v1/ip/geo.json')
+                geoOpn = urllib.request.urlopen(geoReq)
                 geoRes = json.load(geoOpn)
                 geoOpn.close()
-            except urllib2.URLError as urlError:
-                print("Error during geo IP acquisition: " + str(urlError.reason))
+            except urllib.error.URLError as urlError:
+                print(("Error during geo IP acquisition: " + str(urlError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:119]: parenthesis

File: LocastService.py [hunk 12]
```diff
@@ -111,21 +111,21 @@ def validate_user(self):
...
                 return False
-            except urllib2.HTTPError as httpError:
-                print("Error during geo IP acquisition: " + str(httpError.reason))
+            except urllib.error.HTTPError as httpError:
+                print(("Error during geo IP acquisition: " + str(httpError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:122]: parenthesis

File: LocastService.py [hunk 13]
```diff
@@ -111,21 +111,21 @@ def validate_user(self):
...
                 return False
             except:
                 geoIpErr = sys.exc_info()[0]
-                print("Error during geo IP acquisition: " + geoIpErr.message)
+                print(("Error during geo IP acquisition: " + geoIpErr.message))
                 return False
-            print("User location obtained as " + geoRes['latitude'] + '/' + geoRes['longitude'])
+            print(("User location obtained as " + geoRes['latitude'] + '/' + geoRes['longitude']))
```
Comment (deathbybandaid) [inline LocastService.py:128]: parenthesis

File: LocastService.py [hunk 14]
```diff
@@ -136,26 +136,26 @@ def validate_user(self):
...
                                         headers={'Content-Type': 'application/json'})
 
-                dmaOpn = urllib2.urlopen(dmaReq)
+                dmaOpn = urllib.request.urlopen(dmaReq)
                 dmaRes = json.load(dmaOpn)
                 dmaOpn.close()
-            except urllib2.URLError as urlError:
-                print("Error when getting the users's DMA: " + str(urlError.reason))
+            except urllib.error.URLError as urlError:
+                print(("Error when getting the users's DMA: " + str(urlError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:148]: parenthesis

File: LocastService.py [hunk 15]
```diff
@@ -136,26 +136,26 @@ def validate_user(self):
...
                 return False
-            except urllib2.HTTPError as httpError:
-                print("Error when getting the users's DMA: " + str(httpError.reason))
+            except urllib.error.HTTPError as httpError:
+                print(("Error when getting the users's DMA: " + str(httpError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:151]: parenthesis

File: LocastService.py [hunk 16]
```diff
@@ -136,26 +136,26 @@ def validate_user(self):
...
                 return False
             except:
                 dmaErr = sys.exc_info()[0]
-                print("Error when getting the users's DMA: " + dmaErr.message)
+                print(("Error when getting the users's DMA: " + dmaErr.message))
```
Comment (deathbybandaid) [inline LocastService.py:155]: parenthesis

File: LocastService.py [hunk 17]
```diff
@@ -136,26 +136,26 @@ def validate_user(self):
...
                 return False
 
-            print("DMA found as " + dmaRes['DMA'] + ": " + dmaRes['name'])
+            print(("DMA found as " + dmaRes['DMA'] + ": " + dmaRes['name']))
```
Comment (deathbybandaid) [inline LocastService.py:158]: parenthesis

File: LocastService.py [hunk 18]
```diff
@@ -180,23 +180,23 @@ def get_stations(self):
...
 
-        except urllib2.URLError as urlError:
-            print("Error when getting the list of stations: " + str(urlError.reason))
+        except urllib.error.URLError as urlError:
+            print(("Error when getting the list of stations: " + str(urlError.reason)))
             return False
-        except urllib2.HTTPError as httpError:
-            print("Error when getting the list of stations: " + str(httpError.reason))
+        except urllib.error.HTTPError as httpError:
+            print(("Error when getting the list of stations: " + str(httpError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:195]: parenthesis

File: LocastService.py [hunk 19]
```diff
@@ -408,30 +408,30 @@ def find_fcc_station(self, callsign, market, fcc_stations):
 
 
     def get_station_stream_uri(self, station_id):
-        print("Getting station info for " + station_id + "...")
+        print(("Getting station info for " + station_id + "..."))
```
Comment (deathbybandaid) [inline LocastService.py:411]: parenthesis

File: LocastService.py [hunk 20]
```diff
@@ -408,30 +408,30 @@ def find_fcc_station(self, callsign, market, fcc_stations):
...
                                             headers={'Content-Type': 'application/json',
                                                     'authorization': 'Bearer ' + self.current_token})
-            videoUrlOpn = urllib2.urlopen(videoUrlReq)
+            videoUrlOpn = urllib.request.urlopen(videoUrlReq)
             videoUrlRes = json.load(videoUrlOpn)
             videoUrlOpn.close()
-        except urllib2.URLError as urlError:
-            print("Error when getting the video URL: " + str(urlError.reason))
+        except urllib.error.URLError as urlError:
+            print(("Error when getting the video URL: " + str(urlError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:424]: parenthesis

File: LocastService.py [hunk 21]
```diff
@@ -408,30 +408,30 @@ def find_fcc_station(self, callsign, market, fcc_stations):
...
             return False
-        except urllib2.HTTPError as httpError:
-            print("Error when getting the video URL: " + str(httpError.reason))
+        except urllib.error.HTTPError as httpError:
+            print(("Error when getting the video URL: " + str(httpError.reason)))
```
Comment (deathbybandaid) [inline LocastService.py:427]: parenthesis

File: LocastService.py [hunk 22]
```diff
@@ -408,30 +408,30 @@ def find_fcc_station(self, callsign, market, fcc_stations):
...
             return False
         except:
             videoUrlReqErr = sys.exc_info()[0]
-            print("Error when getting the video URL: " + videoUrlReqErr.message)
+            print(("Error when getting the video URL: " + videoUrlReqErr.message))
```
Comment (deathbybandaid) [inline LocastService.py:431]: parenthesis

File: LocastService.py [hunk 23]
```diff
@@ -408,30 +408,30 @@ def find_fcc_station(self, callsign, market, fcc_stations):
...
             return False
 
-        print("Determining best video stream for " + station_id + "...")
+        print(("Determining best video stream for " + station_id + "..."))
```
Comment (deathbybandaid) [inline LocastService.py:434]: parenthesis

File: LocastService.py [hunk 24]
```diff
@@ -440,7 +440,7 @@ def get_station_stream_uri(self, station_id):
 
         
 
-        print("Found " + str(len(videoUrlM3u.playlists)) + " Playlists")
+        print(("Found " + str(len(videoUrlM3u.playlists)) + " Playlists"))
```
Comment (deathbybandaid) [inline LocastService.py:443]: parenthesis

File: LocastService.py [hunk 25]
```diff
@@ -458,9 +458,9 @@ def get_station_stream_uri(self, station_id):
         
 
             if bestStream != None:
-                print(station_id + " will use " + 
+                print((station_id + " will use " +
```
Comment (deathbybandaid) [inline LocastService.py:461]: parenthesis

File: main.py [hunk 26]
```diff
@@ -143,7 +143,7 @@ def do_GET(self):
             self.end_headers()
 
         else:
-            print("Unknown request to " + self.path)
+            print(("Unknown request to " + self.path))
```
Comment (deathbybandaid) [inline main.py:146]: parenthesis

File: main.py [hunk 27]
```diff
@@ -194,14 +194,14 @@ def do_POST(self):
                 self.send_header('Content-type','text/html')
                 self.end_headers()
             else:
-                print("Unknown scan command " + queryData['scan'])
+                print(("Unknown scan command " + queryData['scan']))
```
Comment (deathbybandaid) [inline main.py:197]: parenthesis

File: main.py [hunk 28]
```diff
@@ -194,14 +194,14 @@ def do_POST(self):
...
                 self.send_response(400)
                 self.send_header('Content-type','text/html')
                 self.end_headers()
                 self.wfile.write(self.templates['htmlError'].format(queryData['scan'] + ' is not a valid scan command'))
 
         else:
-            print("Unknown request to " + contentPath)
+            print(("Unknown request to " + contentPath))
```
Comment (deathbybandaid) [inline main.py:204]: parenthesis

File: main.py [hunk 29]
```diff
@@ -318,9 +318,9 @@ def ssdpServerProcess(address, port, uuid):
     REPORTING_FIRMWARE_VER = config["reporting_firmware_ver"]
 
 
-    print("Locast2Plex v" + CURRENT_VERSION)
+    print(("Locast2Plex v" + CURRENT_VERSION))
```
Comment (deathbybandaid) [inline main.py:321]: parenthesis

File: main.py [hunk 30]
```diff
@@ -318,9 +318,9 @@ def ssdpServerProcess(address, port, uuid):
...
 
-    print("Tuner count set to " + str(TUNER_COUNT))
+    print(("Tuner count set to " + str(TUNER_COUNT)))
```
Comment (deathbybandaid) [inline main.py:323]: parenthesis

File: main.py [hunk 31]
```diff
@@ -333,7 +333,7 @@ def ssdpServerProcess(address, port, uuid):
             config_handler.write(config_file)
 
 
-    print("UUID set to: " + DEVICE_UUID + "...")
+    print(("UUID set to: " + DEVICE_UUID + "..."))
```
Comment (deathbybandaid) [inline main.py:336]: parenthesis

File: main.py [hunk 32]
```diff
@@ -357,7 +357,7 @@ def ssdpServerProcess(address, port, uuid):
         station_list = locast.get_stations()
 
         try:
-            print("Starting device server on " + config['plex_accessible_ip'] + ":" + config['plex_accessible_port'])
+            print(("Starting device server on " + config['plex_accessible_ip'] + ":" + config['plex_accessible_port']))
```
Comment (deathbybandaid) [inline main.py:360]: parenthesis

File: m3u8/iso8601/iso8601.py [hunk 33]
```diff
@@ -19,7 +19,7 @@
 if sys.version_info >= (3, 0, 0):
```
Comment (deathbybandaid) [inline m3u8/iso8601/iso8601.py:19]: Py2 is dead, we don't need to include backwards compatibility

File: m3u8/mixins.py [hunk 34]
```diff
@@ -3,7 +3,7 @@
 from m3u8.parser import is_url
 
 try:
```
Comment (deathbybandaid) [inline m3u8/mixins.py:5]: I'd say kill this `as url_parser` stuff and just use `urllib.parse` in the actual code later

**Comment by deathbybandaid** [comment]
You did some really good work here.

**Comment by ratherDashing** [comment]
@deathbybandaid thank you!

**Comment by deathbybandaid** [comment]
If this gets merged before my linting PR, I'd be fine with that @tgorgdotcom 
I'll just rebase my branch.

I'd say this, and the linting, and then a release.

**Comment by deathbybandaid** [comment]
@ratherDashing I noticed you did this PR from your master branch, normally the best way to git collab is: 

1) Fork
2) `git remote add upstream https://github.com/ORIGINAL_OWNER/ORIGINAL_REPOSITORY.git`
3) create a new branch for the feature that you are dev-ing

3.5) If you find yourself with tons of commits, due to testing
     a) `git checkout branchname_to_update`
     b) `git rebase -i master`
     c) `git push -f origin branchname_to_update`

This will then make it look like you are a master coder, that makes fewer commits for a single PR.

4) PR from branch, never master


Then later, keep your fork's master branch up to date with upstream,
1) `git checkout master`
2) `git fetch upstream`
3) `git merge upstream/master`

File: main.py [hunk 35]
```diff
@@ -236,7 +241,7 @@ def __init__(self, serverSocket, config, templates, station_list, locast_service
         self.start()
 
     def run(self):
-        httpd = HTTPServer((self.address, int(self.port)), PlexHttpHandler, False)
```
Comment (deathbybandaid) [inline main.py:239]: removing `False` causes my instance to not be able to bind to the port

**Comment by deathbybandaid** [comment]
I was finding some encoding issues without adding `.encode('utf-8')` to all the lines using `self.wfile.write` as below:

`self.wfile.write(self.templates['jsonLineupComplete'].encode('utf-8'))`

**Comment by deathbybandaid** [comment]
To clarify, not every time `wfile.write`, depends on if it's a `str` or a bytes-like object like video

**Comment by tgorgdotcom** [comment]
I merged this into another branch (so hopefully you still get credit).  I'll use @deathbybandaid's version for main.  Thanks for your contribution!

**Pull request closed without merging**